pkg/apis/autoscaling/v2alpha1/defaults_test.go

```diff
@@ -0,0 +1,131 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v2alpha1_test
+
+import (
+	"reflect"
+	"testing"
+
+	autoscalingv2alpha1 "k8s.io/api/autoscaling/v2alpha1"
+	"k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kubernetes/pkg/api"
+	_ "k8s.io/kubernetes/pkg/api/install"
+	"k8s.io/kubernetes/pkg/apis/autoscaling"
+	_ "k8s.io/kubernetes/pkg/apis/autoscaling/install"
+	. "k8s.io/kubernetes/pkg/apis/autoscaling/v2alpha1"
+)
+
+func TestSetDefaultHPA(t *testing.T) {
+	utilizationDefaultVal := int32(autoscaling.DefaultCPUUtilization)
+	defaultReplicas := newInt32(1)
+	defaultTemplate := []autoscalingv2alpha1.MetricSpec{
+		{
+			Type: autoscalingv2alpha1.ResourceMetricSourceType,
+			Resource: &autoscalingv2alpha1.ResourceMetricSource{
+				Name: v1.ResourceCPU,
+				TargetAverageUtilization: &utilizationDefaultVal,
+			},
+		},
+	}
+
+	tests := []struct {
+		original *autoscalingv2alpha1.HorizontalPodAutoscaler
+		expected *autoscalingv2alpha1.HorizontalPodAutoscaler
+	}{
+		{ // MinReplicas default value
+			original: &autoscalingv2alpha1.HorizontalPodAutoscaler{
+				Spec: autoscalingv2alpha1.HorizontalPodAutoscalerSpec{
+					Metrics: defaultTemplate,
+				},
+			},
+			expected: &autoscalingv2alpha1.HorizontalPodAutoscaler{
+				Spec: autoscalingv2alpha1.HorizontalPodAutoscalerSpec{
+					MinReplicas: defaultReplicas,
+					Metrics:     defaultTemplate,
+				},
+			},
+		},
+		{ // MinReplicas update
+			original: &autoscalingv2alpha1.HorizontalPodAutoscaler{
+				Spec: autoscalingv2alpha1.HorizontalPodAutoscalerSpec{
+					MinReplicas: newInt32(3),
+					Metrics:     defaultTemplate,
+				},
+			},
+			expected: &autoscalingv2alpha1.HorizontalPodAutoscaler{
+				Spec: autoscalingv2alpha1.HorizontalPodAutoscalerSpec{
+					MinReplicas: newInt32(3),
+					Metrics:     defaultTemplate,
+				},
+			},
+		},
+		{ // Metrics default value
+			original: &autoscalingv2alpha1.HorizontalPodAutoscaler{
+				Spec: autoscalingv2alpha1.HorizontalPodAutoscalerSpec{
+					MinReplicas: defaultReplicas,
+				},
+			},
+			expected: &autoscalingv2alpha1.HorizontalPodAutoscaler{
+				Spec: autoscalingv2alpha1.HorizontalPodAutoscalerSpec{
+					MinReplicas: defaultReplicas,
+					Metrics:     defaultTemplate,
+				},
+			},
+		},
+	}
+
+	for i, test := range tests {
+		original := test.original
+		expected := test.expected
+		obj2 := roundTrip(t, runtime.Object(original))
+		got, ok := obj2.(*autoscalingv2alpha1.HorizontalPodAutoscaler)
+		if !ok {
+			t.Fatalf("(%d) unexpected object: %v", i, got)
```

Comment:
Thanks for review, **1)** it should print obj2 , will fix it. **2)**Fatalf has same effect as Errorf and FailNow, only Errorf will continue execution.